File: entities/Class.py
from math import e
from tkinter import SEL
from database import *
from entities.Equipment import *
from entities.Multiclass import *
from entities.Spell import *
from entities.Feature import *
from entities.Choice import *
from entities.Slots import *
from utilities.lists import *

for item in classes:
    class_index = item["index"]
    profs = []    
    proficiencyOptions = [i for i in item["proficiency_choices"][0]["from"]["options"]]

    for options in proficiencyOptions:
        profs.append(options["item"]["index"])
                        
    proficiencyChoices[class_index] = profs

class Class:
    def __init__(self, classe):
        self.name = classe["index"]
        self.hp = classe["hit_die"]
        self.subclasses = [item["index"] for item in classe["subclasses"]]
        self.proficiencyChoices = [Choice(classe["proficiency_choices"][0]["choose"], proficiencyChoices[classe["index"]])]
        self.savingThrows = [item["index"] for item in classe["saving_throws"]]
        self.proficiencies = [item["index"] for item in classe["proficiencies"]]
        self.features = [feature for feature in featureStats if feature.classe == self.name]
        self.startingEquipments = [Equipment(item["equipment"]["index"], item["equipment"]["quantity"]) for item in classe["starting_equipment"] if item["equipment"]["index"] in weapons]
        self.startingEquipmentsOptions = self.getEquipmentOptions()
        self.multiclassing = Multiclass(classe["multi_classing"])
        self.spellcastingAbility = classe["spellcasting"]["spellcasting_ability"]["index"] if "spellcasting" in classe else None
        self.spells = [item for item in spellStats if self.name in item.classes]

    def getEquipmentOptions(self):
        equipmentOptions = []
        choices = []

        if self.name == "barbarian":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("greataxe", 1))
            else:
                for item in martialMeleeWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("handaxe", 2))
            else:
                for item in simpleWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

        if self.name == "bard":
            selection = random.randint(0, 2)
            if selection == 0:
                equipmentOptions.append(Equipment("rapier", 1))
            elif selection == 1:
                equipmentOptions.append(Equipment("longsword", 1))
            else:
                for item in simpleWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

        if self.name == "cleric":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("mace", 1))
            else:
                equipmentOptions.append(Equipment("warhammer", 1))

            selection = random.randint(0, 2)
            if selection == 0:
                equipmentOptions.append(Equipment("scale-mail", 1))
            elif selection == 1:
                equipmentOptions.append(Equipment("leather-armor", 1))
            else:
                equipmentOptions.append(Equipment("chain-mail", 1))

            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("light-crossbow", 1))
                equipmentOptions.append(Equipment("crossbow-bolt", 20))
            else:
                for item in simpleWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

        if self.name == "druid":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("wooden-shield", 1))
            else:
                for item in simpleWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("scimitar", 1))
            else:
                for item in simpleMeleeWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

        if self.name == "fighter":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("chain-mail", 1))
            else:
                equipmentOptions.append(Equipment("leather-armor", 1))
                equipmentOptions.append(Equipment("longbow", 1))
                equipmentOptions.append(Equipment("arrow", 20))

            selection = random.randint(0, 1)
            if selection == 0:
                for item in martialWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)
                equipmentOptions.append(Equipment("shield", 1))
            else:
                for item in martialWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 2)

            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("light-crossbow", 1))
                equipmentOptions.append(Equipment("crossbow-bolt", 20))
            else:
                equipmentOptions.append(Equipment("handaxe", 2))

        if self.name == "monk":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("shortsword", 1))
            else:
                for item in simpleWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

        if self.name == "paladin":
            selection = random.randint(0, 1)
            if selection == 0:
                for item in martialWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)
                equipmentOptions.append(Equipment("shield", 1))
            else:
                for item in martialWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 2)

            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("javelin", 5))
            else:
                for item in simpleMeleeWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

        if self.name == "ranger":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("scale-mail", 1))
            else:
                equipmentOptions.append(Equipment("leather-armor", 1))

            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("shortsword", 2))
            else:
                for item in simpleMeleeWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 2)
            
        if self.name == "rogue":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("rapier", 1))
            else:
                equipmentOptions.append(Equipment("shortsword", 1))

            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("shortbow", 1))
                equipmentOptions.append(Equipment("arrow", 20))
            else:
                equipmentOptions.append(Equipment("shortsword", 1))

        if self.name == "sorcerer":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("light-crossbow", 1))
                equipmentOptions.append(Equipment("crossbow-bolt", 20))
            else:
                for item in simpleWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

        if self.name == "warlock":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("light-crossbow", 1))
                equipmentOptions.append(Equipment("crossbow-bolt", 20))
            else:
                for item in simpleWeapons:
                    choices.append(Choice(Equipment(item["index"], 1)), 1)

            for item in simpleWeapons:
                choices.append(Choice(Equipment(item["index"], 1)), 1)

        if self.name == "wizard":
            selection = random.randint(0, 1)
            if selection == 0:
                equipmentOptions.append(Equipment("quarterstaff", 1))
            else:
                equipmentOptions.append(Equipment("dagger", 1))

        if len(choices) > 0:
            for choice in choices:
                equipmentOptions.extend(choice.getRandomChoice(equipmentOptions))

        return equipmentOptions

    def getSpellSlots(self, level):
        spellSlots = Slots()

        if self.name in ["bard", "cleric", "druid", "sorcerer", "wizard"]:
            if level == 1:
                spellSlots = Slots(2)
            elif level == 2:
                spellSlots = Slots(3)
            elif level == 3:
                spellSlots = Slots(4, 2)
            elif level == 4:
                spellSlots = Slots(4, 3)
            elif level == 5:
                spellSlots = Slots(4, 3, 2)
            elif level == 6:
                spellSlots = Slots(4, 3, 3)
            elif level == 7:
                spellSlots = Slots(4, 3, 3, 1)
            elif level == 8:
                spellSlots = Slots(4, 3, 3, 2)
            elif level == 9:
                spellSlots = Slots(4, 3, 3, 3, 1)
            elif level == 10:
                spellSlots = Slots(4, 3, 3, 3, 2)
            elif level == 11 or level == 12:
                spellSlots = Slots(4, 3, 3, 3, 2, 1)
            elif level == 13 or level == 14:
                spellSlots = Slots(4, 3, 3, 3, 2, 1, 1)
            elif level == 15 or level == 16:
                spellSlots = Slots(4, 3, 3, 3, 2, 1, 1, 1)
            elif level == 17:
                spellSlots = Slots(4, 3, 3, 3, 2, 1, 1, 1, 1)
            elif level == 18:
                spellSlots = Slots(4, 3, 3, 3, 3, 1, 1, 1, 1)
            elif level == 19:
                spellSlots = Slots(4, 3, 3, 3, 3, 2, 1, 1, 1)
            elif level == 20:
                spellSlots = Slots(4, 3, 3, 3, 3, 2, 2, 1, 1)
        elif self.name in ["paladin", "ranger"]:
            if level == 2:
                spellSlots = Slots(2)
            elif level == 3:
                spellSlots = Slots(3)
            elif level == 4:
                spellSlots = Slots(3)
            elif level == 5 or level == 6:
                spellSlots = Slots(4, 2)
            elif level == 7 or level == 8:
                spellSlots = Slots(4, 3)
            elif level == 9 or level == 10:
                spellSlots = Slots(4, 3, 2)
            elif level == 11 or level == 12:
                spellSlots = Slots(4, 3, 3)
            elif level == 13 or level == 14:
                spellSlots = Slots(4, 3, 3, 1)
            elif level == 15 or level == 16:
                spellSlots = Slots(4, 3, 3, 2)
            elif level == 17 or level == 18:
                spellSlots = Slots(4, 3, 3, 3, 1)
            elif level == 19 or level == 20:
                spellSlots = Slots(4, 3, 3, 3, 2)

        return spellSlots


classStats = [Class(item) for item in classes]

File: TrainDataGen/Entities/Class.cs
using MongoDB.Bson.Serialization.Attributes;
using TrainDataGen.Entities.Mappers;

namespace TrainDataGen.Entities;

public class Class
{
    public string Name { get; set; }
    public short Hp { get; set; }
    public List<BaseMapper> Subclasses { get; set; }
    public List<BaseMapper> SavingThrows { get; set; }
    public List<BaseMapper> Proficiencies { get; set; }
    public List<Feature> Features { get; set; }
    public List<Equipment> Equipment { get; set; }
    public Multiclass Multiclassing { get; set; }
    public string? SpellcastingAbility { get; set; }
    public List<Spell> Spells { get; set; }

    public Class (ClassMapper classMapper, List<Feature> features, List<Spell> spells)
    {
        Name = classMapper.Name;
        Hp = classMapper.Hp;
        Subclasses = classMapper.Subclasses;
        SavingThrows = classMapper.SavingThrows;
        Proficiencies = classMapper.Proficiencies;
        Equipment = classMapper.StartingEquipments;
        StartingEquipmentsOptions = classMapper.StartingEquipmentsOptions;
        Multiclassing = classMapper.Multiclassing;
        SpellcastingAbility = classMapper.SpellcastingAbility;
        Features = features;
        Spells = spells;

        classMapper.ProficiencyChoices.ForEach(choice => Proficiencies.AddRange(choice.GetRandomChoice(Proficiencies)));
    }

    public List<Equipment> GetEquipmentOptions(
            List<dynamic> martialMeleeWeapons,
            List<dynamic> simpleWeapons,
            List<dynamic> simpleMeleeWeapons,
            List<dynamic> martialWeapons)
    {
        var equipmentOptions = new List<Equipment>();
        var choices = new List<Choice>();
        var random = new Random();

        if (Name == "barbarian")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("greataxe", 1));
            }
            else
            {
                var listEquip = martialMeleeWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }

            selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("handaxe", 2));
            }
            else
            {
                var listEquip = simpleWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }
        }

        if (Name == "bard")
        {
            int selection = random.Next(3);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("rapier", 1));
            }
            else if (selection == 1)
            {
                equipmentOptions.Add(new Equipment("longsword", 1));
            }
            else
            {
                var listEquip = simpleWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }
        }

        if (Name == "cleric")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("mace", 1));
            }
            else
            {
                equipmentOptions.Add(new Equipment("warhammer", 1));
            }

            selection = random.Next(3);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("scale-mail", 1));
            }
            else if (selection == 1)
            {
                equipmentOptions.Add(new Equipment("leather-armor", 1));
            }
            else
            {
                equipmentOptions.Add(new Equipment("chain-mail", 1));
            }

            selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("light-crossbow", 1));
                equipmentOptions.Add(new Equipment("crossbow-bolt", 20));
            }
            else
            {
                var listEquip = simpleWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }
        }

        if (Name == "druid")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("wooden-shield", 1));
            }
            else
            {
                var listEquip = simpleWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }

            selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("scimitar", 1));
            }
            else
            {
                var listEquip = simpleMeleeWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }
        }

        if (Name == "fighter")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("chain-mail", 1));
            }
            else
            {
                equipmentOptions.Add(new Equipment("leather-armor", 1));
                equipmentOptions.Add(new Equipment("longbow", 1));
                equipmentOptions.Add(new Equipment("arrow", 20));
            }

            selection = random.Next(2);
            if (selection == 0)
            {
                var listEquip = martialWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
                equipmentOptions.Add(new Equipment("shield", 1));
            }
            else
            {
                var listEquip = martialWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(2, listEquip));
            }

            selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("light-crossbow", 1));
                equipmentOptions.Add(new Equipment("crossbow-bolt", 20));
            }
            else
            {
                equipmentOptions.Add(new Equipment("handaxe", 2));
            }
        }

        if (Name == "monk")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("shortsword", 1));
            }
            else
            {
                var listEquip = simpleWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }
        }

        if (Name == "paladin")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                var listEquip = martialWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
                equipmentOptions.Add(new Equipment("shield", 1));
            }
            else
            {
                var listEquip = martialWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(2, listEquip));
            }

            selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("javelin", 5));
            }
            else
            {
                var listEquip = simpleMeleeWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }
        }

        if (Name == "ranger")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("scale-mail", 1));
            }
            else
            {
                equipmentOptions.Add(new Equipment("leather-armor", 1));
            }

            selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("shortsword", 2));
            }
            else
            {
                var listEquip = simpleMeleeWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(2, listEquip));
            }
        }

        if (Name == "rogue")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("rapier", 1));
            }
            else
            {
                equipmentOptions.Add(new Equipment("shortsword", 1));
            }

            selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("shortbow", 1));
                equipmentOptions.Add(new Equipment("arrow", 20));
            }
            else
            {
                equipmentOptions.Add(new Equipment("shortsword", 1));
            }
        }

        if (Name == "sorcerer")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("light-crossbow", 1));
                equipmentOptions.Add(new Equipment("crossbow-bolt", 20));
            }
            else
            {
                var listEquip = simpleWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }
        }

        if (Name == "warlock")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("light-crossbow", 1));
                equipmentOptions.Add(new Equipment("crossbow-bolt", 20));
            }
            else
            {
                var listEquip = simpleWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
                choices.Add(new Choice(1, listEquip));
            }

            var listEquip2 = simpleWeapons.Select(item => new Equipment((string)item.index, 1)).ToList();
            choices.Add(new Choice(1, listEquip2));
        }

        if (Name == "wizard")
        {
            int selection = random.Next(2);
            if (selection == 0)
            {
                equipmentOptions.Add(new Equipment("quarterstaff", 1));
            }
            else
            {
                equipmentOptions.Add(new Equipment("dagger", 1));
            }
        }

        // Add choices to equipmentOptions
        if (choices.Count > 0)
        {
            foreach (var choice in choices)
            {
                equipmentOptions.AddRange(choice.GetRandomChoiceWithoutCheck());
            }
        }

        return equipmentOptions;
    }

    public Slots GetSpellSlots(int level)
    {
        var spellSlots = new Slots();

        if (new[] { "bard", "cleric", "druid", "sorcerer", "wizard" }.Contains(Name))
        {
            spellSlots = level switch
            {
                1 => new Slots(2),
                2 => new Slots(3),
                3 => new Slots(4, 2),
                4 => new Slots(4, 3),
                5 => new Slots(4, 3, 2),
                6 => new Slots(4, 3, 3),
                7 => new Slots(4, 3, 3, 1),
                8 => new Slots(4, 3, 3, 2),
                9 => new Slots(4, 3, 3, 3, 1),
                10 => new Slots(4, 3, 3, 3, 2),
                11 or 12 => new Slots(4, 3, 3, 3, 2, 1),
                13 or 14 => new Slots(4, 3, 3, 3, 2, 1, 1),
                15 or 16 => new Slots(4, 3, 3, 3, 2, 1, 1, 1),
                17 => new Slots(4, 3, 3, 3, 2, 1, 1, 1, 1),
                18 => new Slots(4, 3, 3, 3, 3, 1, 1, 1, 1),
                19 => new Slots(4, 3, 3, 3, 3, 2, 1, 1, 1),
                20 => new Slots(4, 3, 3, 3, 3, 2, 2, 1, 1),
                _ => new Slots()
            };
        }
        else if (new[] { "paladin", "ranger" }.Contains(Name))
        {
            spellSlots = level switch
            {
                2 => new Slots(2),
                3 or 4 => new Slots(3),
                5 or 6 => new Slots(4, 2),
                7 or 8 => new Slots(4, 3),
                9 or 10 => new Slots(4, 3, 2),
                11 or 12 => new Slots(4, 3, 3),
                13 or 14 => new Slots(4, 3, 3, 1),
                15 or 16 => new Slots(4, 3, 3, 2),
                17 or 18 => new Slots(4, 3, 3, 3, 1),
                19 or 20 => new Slots(4, 3, 3, 3, 2),
                _ => new Slots()
            };
        }

        return spellSlots;
    }
}